### ISSUE TITLE
Pages, Page Tree and Navigation

### ISSUE THREAD

**Issue by MartinArendt** [keystonejs/keystone#428]
I did some research for Content Management Systems based on node,js. My favourite would have been nodize but the apparent lack of continuous development was a show-stopper.

So Keystone got on the shortlist. No full-grown CMS compared to what I'm used to in the PHP world though a lightweight design and rather flexible.

Just one downer: it does not support tree structures for pages probably owing to nosql mongodb. I may be antiquated but I consider this as real disadvantage because trees are quite usefull for some purposes like navigation structures, breadcrumbs etc. Is there anyone here who has tried to extend keysone to manage a page tree? Is it a good idea, anyhow?

**Comment by hatzipanis** [comment]
@danielmahon Interested to hear if this is still what you're doing or whether you have refined your process even further. I'm about to implement some kind of Page model and I haven't seen anything more detailed floating around than the above.

**Comment by radiovisual** [comment]
@danielmahon I would also be curious to learn if your approach has been battle-tested. I am about to venture into creating custom Pages from keystone for an upcoming CMS project, and I need something similar to what you have implemented. 

@JedWatson , is it possible to get an entry into the Knowledge base (or a tutorial or "best practices" guide) on the best way to implement a "Pages" feature like the one discussed here? Many of the clients I build for are not technically-inclined, so they need a simple way to create new content on their websites. I am currently evaluating keystone for a massive project, and this is an important feature. *Also, thanks for all the great work you have put into this project!* :+1:

**Comment by LorbusChris** [comment]
+1 to this. I'd really like to use react-router for this. I've been mingling with react-intl (the v2 preprelease) and the keystone-test-project and was actually hoping to build smth like a navbar component with a select language dropdown. Still wrapping my head around server side parsing of localization files and all that, though. Also noticed there was no predefined pages model. I guess thats probably because a page might now just be a component with more components in it. Because I'm pretty new to this I'd like to keep my dev stack a small ap. Thats why I want to avoid templating engines, and I dont think they are needed anymore when using react. This is of course probably means building the public UI also entirely in react. Whats the plan at thinkmill for that?

Tl;dr 
I'd really like to see a sample implementation of a react-router powered navbar for react Pages/Components (whatever you want to call them)

**Comment by JedWatson** [comment]
Dynamic building of navigation actually isn't _too_ challenging in Express; you just need to replace the build-in router for the dynamic routes with your own middleware, and have it look up the path in the database (can be cached for speed if necessary but that's unlikely)

i.e if you wanted to load a page from a `Pages` collection, where each `Page` stores a custom URL path and a `template` file to use to render, it would look something like this:

In your `routes/index.js`:
```
// note: this goes after other routes with dynamic paths, like the blog
app.get(':page', function(req, res, next) {
  Page.findOne().where('path', req.params.page).exec(function(err, page) {
    if (err || !page) return next(err);
    res.render('pages/' + page.template, { page: page });
  });
});
```

Anything you put on the `Page` model would then be available under the `page` variable in the template. The middleware above is template engine agnostic, and could be a smarter if you wanted (e.g. you could predefine datasets to load in checkboxes in the `Page` model, and conditionally run the queries before rendering the template)

You'd probably also store some navigation properties in the `Page` model, like `showInPrimaryNavigation: boolean` that you could use in the templates to look through the items & dynamically generate HTML.

@LorbusChris we're experimenting heavily with Keystone + React + react-router powered sites at Thinkmill at the moment. Generally this means generating whatever page-init data is required into a `<script>` tag when the base template is loaded, and using that to generate whatever we need in React; other data is loaded by the API. We're not really seeing a lot of downside to site front-ends being generated entirely in React these days (especially with options to pre-render HTML server-side available if you need them)

I'm really interested in seeing something developed in this area around pre-build React components that work with a pre-built Keystone model (or convention for configuring Page models)

If you're up for contributing some patterns to the test site (or keystone-react generator) please feel free to propose them - would be good to get something more self-documenting out there.

Right now developments in this area (new react-router about to be released, redux, relay + graphql, etc) mean things are rapidly evolving and we basically have infinite options on how to solve problems... and don't even get me started on style preprocessors vs. styles in components ...

Gotta jump back into code right now - really pushing to get 0.4 released - but I hope this is useful, and feel free to ask me any more specific questions about the ideas above!

**Comment by LorbusChris** [comment]
Thanks @JedWatson, my learning curve is really gaining some momentum now! :)

I actually converted the react-router sidebar example to smth of a navbar. Gonna mash it up with the react-intl translations example. And I will also think of a keystone model for pages, properties might be:
`showInPrimaryNav: boolean`
`ownsSecondaryNav: boolean`
`showInFooterNav: boolean`
or should the Secondary Nav (being a sub nav/sidebar on one page) not be in here?

React Intl lets you declare default formatted messages in the code, and then extracts those messages with a babel plugin. Components can be exported with a HOC like this
```
import {intlShape, injectIntl, defineMessages} from 'react-intl';

class LocalesMenu extends Component {
    render() {
        const {formatMessage} = this.props.intl;

        return (
           <menu>                 
               <li>
                    <a
                        href="/?locale=en-US"
                        title={formatMessage(messages.enUSDescription)}
                    >
                        en-US
                    </a>
               </li>
           </menu>
        );
    }
}

LocalesMenu.propTypes = {
    intl: intlShape.isRequired,
};

export default injectIntl(LocalesMenu);
```
and then rendered like this (client):
```
import {IntlProvider} from 'react-intl';
import App from './components/app';

const {locale, messages} = window.App;

ReactDOM.render(
    <IntlProvider locale={locale} messages={messages}>
        <App />
    </IntlProvider>,
    document.getElementById('container')
);
```
In the provided example Messages are aggregated in one file for translation, and on runtime parsed and injected as needed. Express also sets (lets) default values for locale and messages here:
https://github.com/yahoo/react-intl/blob/master/examples/translations/src/server/index.js

Overall, react-intl works with thes types:
https://github.com/yahoo/react-intl/blob/master/src/types.js

I would love to see these types implemented with keystone. Also formatting of numbers, dates, plurals and relative numbers can be done with it..not sure how much work that would be though. I'll focus on my intl-navbar and get back to you guys once I have something I can confidently show off (in a few days hopefully :)

**Comment by shanepisko** [comment]
I am working on implementing the steps outlined by @danielmahon above and am receiving a few errors, the current one i'm dealing with is
    55|                                                         each link in navigation
    56|                                                                 li.dropdown(class=(section == link.key ? 'active' : null)): a(href=link.href)= link.label
  > 57|                                                                 if link.children.length
    58|                                                                         ul.dropdown-menu
    59|                                                                                 each child in link.children
    60|                                                                                         li: a(href=child.href)= child.label

Cannot read property 'length' of undefined

it is referring to the .children reference of the link, also i am using jade templating

**Comment by danielmahon** [comment]
@shanepisko while I havent touched my above code for a bit, sounds like a simple matter of `link.children` not being set. Add some `console.log` statments before line 57 and see if/what `link.children` is being assigned. You may need to append line 57 to be `if link.children && link.children.length`

**Comment by shanepisko** [comment]
Thanks for getting back! i know this is a bit of a dead issue. I will try that. I am now trying to implement this  using relationships and referencing other pages as i was having an issue of other pages showing up using the plucking of the keystone.get('navigation')

**Comment by mxstbr** [comment]
The question seems to have been answered, so I'll close this issue for now. Let me know if you still need something here by commenting!

**Comment by benjaminapetersen** [comment]
New to keystone, I see this issue is closed of 2 years, but not sure it currently supports dynamic pages & menus via CMS/admin? Just getting into the docs, haven't downloaded to give it a go yet. Def hoping there is a solution, or to know what the current workaround is. Thx!

**Comment by danielmahon** [comment]
@benjaminapetersen there is definitely a way to handle dynamic page routing and menu with keystone. You still have to build the routing logic (see above) to tell it how to handle your paths. This part is not handled by the admin UI. You can think of the admin UI as a convenient database interface for your models. The basic premise of my example code is still the same for the latest version (that is just one way to handle it). Create a route that matches your dynamic path. Use the route params (parent/slug) to find the correct page in your database and then return that page’s content in a shared template so you don’t need to create a template for each result. This is a very common use case. It’s bascially the same as the blog/article example in the keystone boilerplate. Just instead of looking up an article’s record and displaying it you are looking up a page. I tend to break a page up into sections. So I would have a ‘page’ model that can hold many ‘section’ models. That way I can easily reuse sections (like calls to action) on different pages.  Once the structure is setup it is easy to just create new pages and sections in the admin UI and you don’t need to mess with the routing logic anymore.

**Comment by benjaminapetersen** [comment]
Ok great, thx! I'll give that another look above. Sounds like a common enough use case it could be a handy customization tutorial!

**Comment by jelordreygulle** [comment]
@danielmahon  , Cannot read property '0' of undefined , 
parent: { type: Types.Select, options: _.pluck(keystone.get('navigation'), 'key').join(','), required: true, default: keystone.get('navigation')[0].key},

**Comment by jelordreygulle** [comment]
What I did on implementing page tree and navigation is this , this is simple but i think it could be somehow a solution

#Page Model
```
Pages.add({
	name: { type: String, required: true },
	key: { type: String, required: false },
	link_destination : { type: Types.Url, require: true},
	second_level_page: { type: Types.Relationship, ref: 'SecondLevelPage', many: true },
	publishedDate: { type: Types.Date, index: true, dependsOn: { state: 'published' } },
	state: { type: Types.Select, options: 'draft, published, archived', default: 'draft', index: true },
	
});
```
#Second Level Page Model
```
SecondLevelPage.add({
    title: { type: String, required: true },
    link_destination : { type: Types.Url, require: true},
    state: { type: Types.Select, options: 'draft, published, archived', default: 'draft', index: true },
    author: { type: Types.Relationship, ref: 'User', index: true },
    publishedDate: { type: Types.Date, index: true },
    content: {
      html: { type: Types.Html, wysiwyg: true, height: 600 }
    },
    
  });
```

#And then in the views
```

//Load the top level pages including the second level
	view.on('init', function (next) {

		var q = keystone.list('Pages').paginate({
			page: req.query.page || 1,
			perPage: 10,
			maxPages: 10,
			filters: {
				state: 'published',
			},
		})
			.sort('publishedDate')
			.populate('second_level_page')

		q.exec(function (err, results) {
			locals.data.toplevelpages = results;
			console.log("Response Data:", locals.data.toplevelpages)
			next(err);
		});
	});
```
#And then on the template 


								{% for page in data.toplevelpages.results %}

								<li class="{{ 'active' if page.key == section else '' }} dropdown ">
									<a class="" data-href="/"
										href="/{{ page.link_destination }}" id="homePage" title="{{ page.key }}"
										aria-label="Home"
										onclick="TrackNavigationClick('topNav', this);">{{ page.name }} </a>

										<div class="dropdown-content">
											{%for second_level in page.second_level_page %}
											<a href="/{{ second_level.link_destination }}" style="color:white;font-size:12.5px;text-align: left; padding: 12px 16px;text-transform: capitalize ;">{{ second_level.title }}</a>
											{% endfor %}
										</div>
								</li>
								{% endfor %}
